### PR TITLE
Apply redaction to latest_event in RoomInfo when applicable

### PR DESCRIPTION
**Pull request by jplatte**

### CODE REVIEW

**Comment by jplatte** [comment]
Only talk half a dozen tries, now it works 😅

File: crates/matrix-sdk-base/src/sliding_sync.rs [hunk 1]
```diff
@@ -765,6 +767,56 @@ mod test {
...
+        let room = room_with_timeline(&[redaction]);
+        let response = response_with_room(room_id, room).await;
+        client.process_sliding_sync(&response).await.expect("Failed to process sync");
+
+        // Then the room still holds the latest event
+        let client_room = client.get_room(room_id).expect("No room found");
+        let latest_event = client_room.latest_event().unwrap();
+        assert_eq!(latest_event.event_id().unwrap(), "$ida");
+
+        // But it's now redacted
```
Comment (bnjbvr) [inline crates/matrix-sdk-base/src/sliding_sync.rs:811]: Nice literate ~~~programming~~~ testing! Is it worth adding a test for a redaction event that's impossible to parse too, to make sure the latest event is at least removed?

File: crates/matrix-sdk-base/src/rooms/normal.rs [hunk 2]
```diff
@@ -851,9 +852,22 @@ impl RoomInfo {
...
+        event: &OriginalSyncRoomRedactionEvent,
+        _raw: &Raw<OriginalSyncRoomRedactionEvent>,
+    ) {
+        #[cfg(feature = "experimental-sliding-sync")]
+        if let Some(latest_event) = &mut self.latest_event {
+            if latest_event.event_id().as_deref() == Some(&*event.redacts) {
+                let room_version = self.base_info.room_version().unwrap_or(&RoomVersionId::V1);
+                match apply_redaction(&latest_event.event, _raw, room_version) {
+                    Some(redacted) => latest_event.event = redacted,
+                    None => self.latest_event = None,
```
Comment (bnjbvr) [inline crates/matrix-sdk-base/src/rooms/normal.rs:866]: When that happens, do we want to look for an older `latest_event`?

File: crates/matrix-sdk-base/src/sliding_sync.rs [hunk 3]
```diff
@@ -765,6 +767,56 @@ mod test {
         assert_eq!(ev_id(client_room.latest_event()), "$idb");
     }
 
+    #[async_test]
+    async fn cached_latest_event_can_be_redacted() {
```
Comment (bnjbvr) [inline crates/matrix-sdk-base/src/sliding_sync.rs:771]: Most tests for `latest_event` seem to live in `matrix-sdk-base`, would it make sense to move this one there as well?

File: crates/matrix-sdk-base/src/rooms/normal.rs [hunk 4]
```diff
@@ -984,6 +998,41 @@ impl RoomInfo {
...
 
+#[cfg(feature = "experimental-sliding-sync")]
+fn apply_redaction(
+    event: &Raw<AnySyncTimelineEvent>,
+    raw_redaction: &Raw<OriginalSyncRoomRedactionEvent>,
+    room_version: &RoomVersionId,
+) -> Option<Raw<AnySyncTimelineEvent>> {
+    use ruma::canonical_json::redact_in_place;
+
+    let mut event_json = match event.deserialize_as() {
```
Comment (bnjbvr) [inline crates/matrix-sdk-base/src/rooms/normal.rs:1009]: uber-nit: here and below, could use `let else` syntax

**Review by bnjbvr — APPROVED** [review]

File: crates/matrix-sdk-base/src/sliding_sync.rs (hunk 1 above)
Comment (jplatte) [reply crates/matrix-sdk-base/src/sliding_sync.rs:811]: Idk, it seems like a rather obscure edge case.

File: crates/matrix-sdk-base/src/rooms/normal.rs (hunk 2 above)
Comment (jplatte) [reply crates/matrix-sdk-base/src/rooms/normal.rs:866]: I don't think so.. Conceptually it's still redacted. If it weren't stored as a raw event, we'd probably have a `Redacted` variant here that we set the latest event to without attempting to actually apply the redaction properly. Maybe this is a good reason not to use a raw event actually, but I think we should discuss that separately from this PR.

File: crates/matrix-sdk-base/src/sliding_sync.rs (hunk 3 above)
Comment (jplatte) [reply crates/matrix-sdk-base/src/sliding_sync.rs:771]: This *is* `matrix-sdk-base`.

File: crates/matrix-sdk-base/src/rooms/normal.rs (hunk 4 above)
Comment (jplatte) [reply crates/matrix-sdk-base/src/rooms/normal.rs:1009]: I would love it if you were right but unfortunately, `let Ok(…) else` doesn't let you access the error in the else branch.

File: crates/matrix-sdk-base/src/sliding_sync.rs (hunk 3 above)
Comment (bnjbvr) [reply crates/matrix-sdk-base/src/sliding_sync.rs:771]: Please never mind my hallucinations.

File: crates/matrix-sdk-base/src/rooms/normal.rs (hunk 2 above)
Comment (bnjbvr) [reply crates/matrix-sdk-base/src/rooms/normal.rs:866]: Sounds good!

**Pull request merged**